source/opt/pass.cpp

```diff
@@ -141,6 +141,11 @@ uint32_t Pass::GenerateCopy(Instruction* object_to_copy, uint32_t new_type_id,
     }
     return ir_builder.AddCompositeConstruct(new_type_id, element_ids)
         ->result_id();
+  } else if (original_type->AsImage() && new_type->AsImage()) {
+    // Allow storing an image with a different format.
+    // For example in function parameters the format is unknown, but inlining
+    // fixes this
+    return object_to_copy->result_id();
```

Comment:
This seems to be different than what the function is suppose to do.  The description of this function says that the type of the value returned is suppose to be the new type, but in this case it won't be.  We might need to find a different way of doing this.

https://github.com/KhronosGroup/SPIRV-Tools/blob/a61600c7639c1a03df4967806fb8cc48ea3c45ca/source/opt/pass.h#L139

I'm thinking that this change would fit in well with the "FixStorageClass" pass, but I would have to see examples.

Reply:
This change was essentially intended to get rid of the assertion when there is a OpStore <variable> <some_image>, and the type of <variable> and <image> are different because they have different image formats (e.g. Rgba16f for the one that is specified and Rgba32f (default) for the variable that is being assigned to). And in this case I'd like the object to copy overriding the type for the variable, which seems to automatically happen in the later legalization passes, so all that is needed here is to get rid of the assertion at the end of GenerateCopy.

I do realize that this may not be a proper way to do it though, maybe it would be required to propagate the type somehow in this case instead of just ignoring the type mismatch?

Here's an example input SPIR-V with the type mismatch in OpStore (which does not matter as the store is eliminated in some part of the legalization) that causes the assertion fail without this change:

```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 16
; Schema: 0
               OpCapability Shader
               OpCapability SampledBuffer
               OpCapability ImageBuffer
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %main "main"
               OpExecutionMode %main LocalSize 1 1 1
               OpSource HLSL 600
               OpName %type_buffer_image "type.buffer.image"
               OpName %Buf "Buf"
               OpName %main "main"
               OpName %src_main "src.main"
               OpName %bb_entry "bb.entry"
               OpName %type_buffer_image_0 "type.buffer.image"
               OpName %b "b"
               OpDecorate %Buf DescriptorSet 0
               OpDecorate %Buf Binding 0
      %float = OpTypeFloat 32
%type_buffer_image = OpTypeImage %float Buffer 2 0 0 2 Rgba16f
%_ptr_UniformConstant_type_buffer_image = OpTypePointer UniformConstant %type_buffer_image
       %void = OpTypeVoid
          %7 = OpTypeFunction %void
%type_buffer_image_0 = OpTypeImage %float Buffer 2 0 0 2 Rgba32f
%_ptr_Function_type_buffer_image_0 = OpTypePointer Function %type_buffer_image_0
        %Buf = OpVariable %_ptr_UniformConstant_type_buffer_image UniformConstant
       %main = OpFunction %void None %7
          %8 = OpLabel
          %9 = OpFunctionCall %void %src_main
               OpReturn
               OpFunctionEnd
   %src_main = OpFunction %void None %7
   %bb_entry = OpLabel
          %b = OpVariable %_ptr_Function_type_buffer_image_0 Function
         %15 = OpLoad %type_buffer_image %Buf
               OpStore %b %15
               OpReturn
               OpFunctionEnd
```

With this change the output after the fix_storage_class pass is identical (but just doesn't assert) so I can't add a good test for it.

To give some insight, here is the HLSL that would generate this SPIR-V:
```
[[vk::image_format("Rgba16f")]]
RWBuffer<float4> Buf;

[numthreads(1, 1, 1)] void main()
{
    RWBuffer<float4> b = Buf;
}
```

Reply:
Thanks for the example.  I won't be able to look at this right away.  However, I don't think this is in too much of a hurry because it is only needed because of a DCX change that has not landed yet.

My idea now it that we should avoid calling GenerateCopy for this case instead of adding an exception inside the function.

Reply:
@s-perron Thanks for looking into this, I've now changed it the way you described: GenerateCopy is simply not called for OpStore of image types, and it seems to still work for my test cases. I've also added a test, which is in a separate commit so it can be reverted if it's not necessary - all it actually does is that it makes sure that the OpStore of mismatching image types does not break.

I'm afraid the DXC side change cannot be merged until this change is integrated into SPIRV-Tools, but I can ask if we could get some confirmation about the change from the DXC side.